### PR TITLE
chore: implement a Java client SDK for Carbonio Mailbox SOAP APIs

### PR DESCRIPTION
**Pull request by matteobaglini**
Hi,

this PR completes the work the past one #358 and includes:
- a "nice" client entry point with builders.
- integration tests with a `mockServer` that simulates the mailbox providing real response dumped via curl.
- a README file with basic instructions and documentations.

### CODE REVIEW

**Review by sonersivri — APPROVED** [review]

**Review by frisonisland — APPROVED** [review]

**Review by keshavbhatt — APPROVED** [review]

**Comment by federicorispo** [comment]
I have some doubts on this implementation that I would like to discuss before I approve this PR:

### SDK project inside carbonio-mailbox repo
Including the http client inside the mailbox project itself could be a counter productive choice for various reasons:
- Being included inside the same carbonio-mailbox pipeline, it could happen that a new version of the SDK is published  for each version of carbonio-mailbox, even if there are no actual changes that require a new version of the SDK
- It is not coherent with the other projects: we were required to create a client SDK (see: storages-sdk, files-sdk, preview-sdk, user-management-sdk, chats-sdk etc). Moreover, the choice of having a sort of monorepo that includes everything that is related to the mailbox would of course force us to have the ui components too (carbonio-mails-ui, carbonio-calendars-ui and carbonio-contacts-ui), since they can also be considered as carbonio-mailbox clients just like the SDK is: this is not doable and not in line with the inner structure of the ui client projects.

One of the reasons that lead you to insert the SDK in the same repo is the genaration of the WSDL from the mailbox: this current PR is, in fact (as stated in the description), the follow-up of [another PR](https://github.com/zextras/carbonio-mailbox/pull/358) that allows the mailbox to generate all the classes that the client requires, starting from the generation of the WSDL (and to clarify, it’s a PR where less than half the reviewers that this current PR has were set as reviewers and it was merged anyway). At a first glance, this could seem like a good enough reason to move the SDK inside of the repo, but in my opinion creating and saving a new package on artifactory called `carbonio-mailbox-schemas` (or similar, it’s a fantasy name) containing the WSDL generated by the mailbox would have been a much cleaner and more tidy solution. Subsequently, the client SDK could add this package as a dependency and it would have the generation of the classes on the inside in a separated pipeline (as it is currently being done by carbonio-user-management).

### Implementation of the “nice” client
When the client acts as a wrapper of the SOAP calls, it should never expose any kind of logic to the user, but just help the user by doing most of the astraction on the inside.
In these examples it is very clear that the user is supposed to build on its own the SOAP call, with the end result that said composition might not be valid.
```java
    final var info = serviceClient.send(ServiceRequests.Info.allSections().withAuthToken(token));

    final var info2 =
        serviceClient.send(
            ServiceRequests.Info.sections("children", "attrs", "prefs").withAuthToken(token));

    final var info3 =
        serviceClient.send(ServiceRequests.Info.sections("children").withAuthToken(token));

    final var accountInfo =
        serviceClient.send(ServiceRequests.AccountInfo.byId(id).withAuthToken(token));

    final var accountInfo2 =
        serviceClient.send(ServiceRequests.AccountInfo.byEmail(email).withAuthToken(token));

    final var domainInfo =
        adminClient.send(AdminServiceRequests.DomainInfo.byName(domain).withAuthToken(tokenAdmin));

    final var domainInfo2 =
        adminClient.send(AdminServiceRequests.Domain.byName(domain).withAuthToken(tokenAdmin));

    final var domainInfo3 =
        adminClient.send(AdminServiceRequests.AccountInfo.byEmail(email).withAuthToken(tokenAdmin));
```

The client is exposing the method “send”: it is such a generic name that it allows the user to call it with the wrong attributes, without the client ever failing at build time. Example:
```java
final var info2 =
        serviceClient.send(
            ServiceRequests.Info.sections("banana", "gigino", ).withAuthToken(token));
```
The behavior that I would expect is to call the method `serviceClient.getInfo(List<InfoSection> sections, String token)` so that both the input and the output of that method are very clear.

### Test
I have noticed that the testing environment has been created by taking inspiration from the[ user-management’s testing environment](https://github.com/zextras/carbonio-user-management/tree/develop/core/src/test): the only things that change are the way in which the requests and the responses have been used.
In the user-management I have created requests and responses that are parametric so that whoever will create more tests in the future will be able to reuse the XML for failing cases, or with different values. In this SDK, instead, the values have been hard-coded making the XMLs not at all flexible nor reusable. If these values were to increase in number, this would mean that we will be going to have a huge number of similar requests and responses files to test the cases of both success and failure of the same API.

**Review by federicorispo — CHANGES_REQUESTED** [review]
See my previous comment

**Comment by frisonisland** [comment]
@federicorispo I just want to put down few lines:
- I think the client was put here just for simplicity: for me (and us as a Team I think) it never worked having to deal with multiple repos because the whole code is too coupled, so a change in one repo meant changing all the others.
- It should be possible to define the client in a different repo as you said, just by putting the soap library as dependency since it provides the wsdl. But still the soap library, as all other mailbox libraries, is bumped every month, and that will cause "carbonio-mailbox-schema" to bump, so at the end the consumers are forced to bump the library every month, because the schema is bumped every month.
Looking forward to discuss together, your concerns make sense to me

**Comment by matteobaglini** [comment]
Hi @federicorispo and @frisonisland, I've integrated all yours feedback. 
Note that, for now, the sdk is still contained in this repo but the idea is to move it in a proper repo (with it's own release cycle) after this PR has been merged.

**Review by federicorispo — APPROVED** [review]
@matteobaglini Everything seems good and follow the given suggestions! Super :rocket:

File: pom.xml [hunk 1]
```diff
@@ -853,7 +848,6 @@
     <module>common</module>
     <module>native</module>
     <module>store-conf</module>
-    <module>mailbox-sdk</module>
```
[the remark this answers is not in the document]
Comment (matteobaglini) [reply pom.xml:856]: More or less :-) in particular it will not deployed until extracted in a different repo.

**Pull request merged**